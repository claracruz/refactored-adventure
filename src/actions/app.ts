import axios from 'axios';
import {
	actionTypes
} from '../constants';
import {
	IActionRequestWeatherData,
	IActionReceiveWeatherData,
	IActionFetchWeatherDataError
} from '../interfaces/weather';
import {
	WEATHER_API
} from '../constants/api';

const requestWeatherData = (text: string) : IActionRequestWeatherData => {
	return {
		type: actionTypes.REQUEST_WEATHER_DATA,
		text
	}
};

const receiveWeatherData = (text: string, data: any) : IActionReceiveWeatherData => {
	return {
		type: actionTypes.RECEIVE_WEATHER_DATA,
		data
	}
};

const weatherDataRequestFailed = (data: any) : IActionFetchWeatherDataError => {
	return {
		type: actionTypes.WEATHER_DATA_REQUEST_FAILED,
		error: data.response.data
	}
};

export const fetchWeatherData = (cityId: string, text: string) => {
	return (dispatch: any) => {
		dispatch(requestWeatherData(text));
		return axios({
			method: 'GET',
			url: WEATHER_API + cityId
		}).then((res: any) =>
			dispatch(receiveWeatherData(text, res.data))
		).catch(function (res: any) {
			dispatch(weatherDataRequestFailed(res))
		});
	}
};
