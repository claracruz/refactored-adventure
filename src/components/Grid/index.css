.grid {
    border-spacing: 0;
    border-collapse: collapse;
}

.grid th,
.grid td {
    padding: 1em 0.6em;
    border: 1px solid #ccc;
    text-align: left;
}

.grid th {
    font-size: .85em;
    background: #e1e1e1
}

.grid th:first-child {
    width: 20%;
}
