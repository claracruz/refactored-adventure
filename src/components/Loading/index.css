.loading {
    position: fixed;
    width: 100%;
    height: 100%;
    top: 0;
    padding: 30% 0;
    background-color: rgba(255, 255, 255, 0.7);
}
