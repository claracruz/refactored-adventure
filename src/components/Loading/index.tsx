import React from 'react';
import './index.css';

export const Loading = () => {
	return (
		<div className="loading">loading...</div>
	);
};
