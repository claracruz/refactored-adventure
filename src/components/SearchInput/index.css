.search-input input {
    width: 230px;
    margin: 0 10px;
}

.search-input button {
    width: 75px;
}
