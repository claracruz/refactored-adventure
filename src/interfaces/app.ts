import { IWeather } from './weather';
import {ISearchComponentProps} from "./search";

export interface IAppComponentState {
	error: string;
	loading: boolean;
	weather: Array<IWeather> | [];
}

export interface IAppState {
	app: IAppComponentState
}

export interface IAppComponentProps extends IAppComponentState, ISearchComponentProps {}
