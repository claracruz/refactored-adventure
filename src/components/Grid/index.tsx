import React from 'react';
import { IWeather } from '../../interfaces/weather';
import { IGridComponentProps } from '../../interfaces/grid';
import './index.css';

export const Grid = ({ className, items } : IGridComponentProps) => {

	return (
		<table className={`${className} grid`}>
			<thead>
			<tr>
				<th>City</th>
				<th>12am</th>
				<th>6am</th>
				<th>12pm</th>
				<th>6pm</th>
			</tr>
			</thead>
			<tbody>{
				items.map((item: IWeather, index: number) => {
					return (
						<tr key={`${item.city}_${index}`}>
							<td>{item.city}</td>
							<td>{item.twelveAM}</td>
							<td>{item.sixAM}</td>
							<td>{item.twelvePM}</td>
							<td>{item.sixPM}</td>
						</tr>
					);
				})
			}</tbody>
		</table>
	);
};
