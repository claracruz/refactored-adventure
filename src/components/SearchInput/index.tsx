import React, {PureComponent} from 'react';
import { ISearchComponentProps } from '../../interfaces/search';
import './index.css';

interface ISearchComponentState {
	searchTerm: string;
}

export class SearchInput extends PureComponent<ISearchComponentProps, ISearchComponentState> {

	constructor(props: ISearchComponentProps) {
		super(props);
		this.state = {
			searchTerm: ''
		}
	}

	onSubmit = () => {
		this.props.onSearch(this.state.searchTerm);
		this.setState({
			searchTerm: ''
		});
	};

	handleChange = (e: React.FormEvent<HTMLInputElement>) => {
		this.setState({
			searchTerm: e.currentTarget.value
		});
	};

	handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
		if (e.key === 'Enter') {
			this.onSubmit();
		}
	};

	render() {
		return (
			<div className="search-input">
				<label>
					<input
						value={this.state.searchTerm}
						onKeyDown={this.handleKeyDown}
						onChange={this.handleChange}
						placeholder="Search a city"
						type="text" />
				</label>
				<button onClick={this.onSubmit}>Search</button>
			</div>
		);
	}
}
