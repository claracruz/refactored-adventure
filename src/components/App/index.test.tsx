import React from 'react';
import { App } from './';
import {IAppComponentProps} from '../../interfaces/app';
import { IGridComponentProps } from '../../interfaces/grid';
import { ISearchComponentProps } from '../../interfaces/search';
import { shallow } from 'enzyme';
import { SearchInput } from '../SearchInput';
import { Grid } from '../Grid';
import { Loading } from '../Loading';

describe('<App />', () => {
	let props : IAppComponentProps;

	beforeEach(() => {
		props = {
			error: '',
			loading: false,
			weather: [],
			onSearch: jest.fn()
		};
	});

	it('renders without crashing', () => {
		const wrapper = shallow(<App {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders without crashing in loading mode', () => {
		props.loading = true;
		const wrapper = shallow(<App {...props} />);
		const loading = <Loading />;
		expect(wrapper.contains(loading)).toEqual(true);
	});

	it('should fetch weather data', () => {
		const wrapper = shallow(<App {...props} />);
		const search = wrapper.find('SearchInput');
		const searchTerm = 'test-call';
		const searchProps = search.props() as ISearchComponentProps;
		searchProps.onSearch(searchTerm);
		expect(props.onSearch).toBeCalledWith(searchTerm);
	});

	it('should render weather data', () => {
		props.weather = [
			{ city: 'city 1', twelveAM: '12C', sixAM: '14C', twelvePM: '89C', sixPM: '44C'},
			{ city: 'city 2', twelveAM: '126C', sixAM: '148C', twelvePM: '899C', sixPM: '4C'},
			{ city: 'city 3', twelveAM: '124C', sixAM: '154C', twelvePM: '819C', sixPM: '54C'}
		];
		const wrapper = shallow(<App {...props} />);
		const grid = wrapper.find('Grid');
		const gridProps = grid.props() as IGridComponentProps;
		expect(gridProps.items).toEqual(props.weather);
	});
});
