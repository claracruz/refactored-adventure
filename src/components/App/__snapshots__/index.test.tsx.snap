// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    error=""
    loading={false}
    onSearch={[MockFunction]}
    weather={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
        >
          <h1>
            24 hour weather forecast
          </h1>
        </header>,
        <section>
          <SearchInput
            onSearch={[Function]}
          />
          <Grid
            className="weather-grid"
            items={Array []}
          />
        </section>,
        false,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1>
            24 hour weather forecast
          </h1>,
          "className": "App-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "24 hour weather forecast",
          },
          "ref": null,
          "rendered": "24 hour weather forecast",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <SearchInput
              onSearch={[Function]}
            />,
            false,
            <Grid
              className="weather-grid"
              items={Array []}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onSearch": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "weather-grid",
              "items": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "section",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
          >
            <h1>
              24 hour weather forecast
            </h1>
          </header>,
          <section>
            <SearchInput
              onSearch={[Function]}
            />
            <Grid
              className="weather-grid"
              items={Array []}
            />
          </section>,
          false,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              24 hour weather forecast
            </h1>,
            "className": "App-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "24 hour weather forecast",
            },
            "ref": null,
            "rendered": "24 hour weather forecast",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SearchInput
                onSearch={[Function]}
              />,
              false,
              <Grid
                className="weather-grid"
                items={Array []}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onSearch": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "weather-grid",
                "items": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
